# CELL 1 -  Final Analysis Report

## Controlled Natural Language Generation
## Decoder Transformer vs Large Language Model (LLM)


## Problem Statement

Build and compare two Natural Language Generation approaches for slot-conditioned response generation.

Approaches:

1. Custom Transformer Decoder Model
2. Pretrained Large Language Model (FLAN-T5)


Input:

Intent + Slots


Example:

intent : order_status

order_id : ORD841324


Expected Output:

I can help with order status for order ORD841324.



Evaluation Focus:

- Response quality
- Slot preservation
- Generalization
- Model limitations

In [1]:
# CELL 2 - Load Evaluation Results


import pandas as pd


evaluation_results = pd.read_csv(

    "../outputs/evaluation_results.csv"

)


evaluation_results

,Model,BLEU,Slot Accuracy,Exact Match
0,Decoder Transformer,0.597315,0.266667,0.066667
1,LLM FLAN-T5,0.110107,0.866667,0.000000


# CELL 3 - Overall Architecture

## Decoder Transformer Flow


Intent + Slots

        |

        v

Linearized Sequence

        |

        v

Tokenization

        |

        v

Embedding Layer

        |

        v

Positional Encoding

        |

        v

Transformer Decoder

        |

        v

Generated Response



---


## LLM Flow


Intent + Slots

        |

        v

Input Cleaning

        |

        v

Prompt Engineering

        |

        v

FLAN-T5 Model

        |

        v

Generated Response


# CELL 4 -  Decoder Transformer Analysis


## Implementation

The custom decoder model was implemented using PyTorch Transformer components.


Major components:

- Token Embedding
- Positional Encoding
- Transformer Decoder Layers
- Linear Prediction Layer


The model was trained using:

Loss Function:

Cross Entropy Loss


Optimizer:

Adam Optimizer



## Observations


The decoder successfully learned:

- Intent based response patterns
- Sentence structure
- Domain specific templates


Example:


Input:

order_id : ORD841324


Expected:

I can help with order status for order ORD841324.


Generated:

I can help with order status for order ORD387515.



## Limitation


Although the sentence structure was correct, the model failed to preserve unseen slot values.


Reason:

- Small training dataset
- Unique IDs rarely repeat
- No explicit copy mechanism

# CELL 5 - LLM Based Generation Analysis

Model Used:

FLAN-T5


Approach:

A pretrained LLM was used with:

- Input preprocessing
- Few-shot prompting
- Controlled instructions



Example:


Input:

order_id : ORD841324


Generated:

I want to order status for order ORD841324



## Observations


The LLM:

- Preserved unseen slot values
- Required no training
- Generated flexible responses



## Limitations


- Output wording can differ from reference
- Prompt design affects quality
- Smaller LLMs may produce grammar variations

# CELL 6 - Evaluation Metrics 

Three metrics were used:


## 1. BLEU Score

Full Form:

Bilingual Evaluation Understudy


Purpose:

Measures word overlap between generated text and reference response.


Higher BLEU means closer wording.



---


## 2. Slot Accuracy


Purpose:

Checks whether important slot values are preserved.


Example:


Expected:

ORD841324


Generated:

ORD841324


Result:

Correct



---


## 3. Exact Match


Purpose:

Checks whether generated response exactly matches the expected sentence.


It is a strict metric because small wording changes cause failure.

# CELL 7 - Actual Result Analysis

# Evaluation Results Analysis


## Decoder Transformer


Results:


BLEU Score:

0.597


Slot Accuracy:

0.267


Exact Match:

0.067



## Interpretation


The decoder achieved higher BLEU score because it learned fixed response patterns from training data.


However:

The low slot accuracy shows that the model generated fluent sentences but struggled with factual slot preservation.



---



## LLM FLAN-T5


Results:


BLEU Score:

0.110


Slot Accuracy:

0.867


Exact Match:

0.000



## Interpretation


The LLM produced different wording compared to reference sentences.


Because BLEU and Exact Match depend on word similarity, the scores are lower.


However, the LLM preserved slot values much better.


This is important for real conversational systems.

# CELL 8 - Comparision Table -> Decoder vs LLM Comparison


| Feature | Decoder Transformer | FLAN-T5 LLM |
|---|---|---|
| Training Required | Yes | No |
| Learns Templates | High | Medium |
| BLEU Score | Higher | Lower |
| Slot Preservation | Lower | Higher |
| Handles New IDs | Weak | Strong |
| Compute Cost | Lower | Higher |
| Prompt Required | No | Yes |
| Flexibility | Lower | Higher |



# CELL 9 - Production Recommendation

For real-world Conversational AI:


Decoder based systems are suitable when:

- Domain is fixed
- Responses are highly controlled
- Low latency is required
- Compute resources are limited



LLM based systems are suitable when:

- Inputs vary frequently
- Better generalization is needed
- Dynamic slot values must be preserved



Recommended Hybrid Architecture:


Intent + Slots

        |

        v

LLM Generation

        |

        v

Slot Validation Layer

        |

        v

Business Rule Verification

        |

        v

Final Response


This combines flexibility with reliability.

# CELL 10 - # Final Conclusion


The experiment demonstrates the trade-off between a custom Transformer Decoder and a pretrained LLM.


The Decoder Transformer:

- Achieved higher BLEU score (0.597)
- Learned response templates effectively
- Generated fluent sentences


However:

- It struggled with unseen slot copying
- Slot accuracy was only 0.267



The FLAN-T5 LLM:

- Achieved higher slot accuracy (0.867)
- Preserved important values better
- Required no model training


However:

- Generated different wording
- Had lower BLEU score



Final Observation:

BLEU alone is not sufficient for evaluating task-oriented dialogue generation.

For slot-conditioned NLG systems, factual correctness and slot preservation are more important.


A hybrid LLM + validation approach provides the best practical solution.